python_modules/dagster/dagster/_core/storage/event_log/base.py

```diff
@@ -411,3 +411,48 @@ def alembic_version(self) -> Optional[AlembicVersion]:
     def is_run_sharded(self) -> bool:
         """Indicates that the EventLogStoarge is sharded."""
         return False
+
+    @property
+    def supports_global_concurrency_limits(self) -> bool:
+        """Indicates that the EventLogStorage supports global concurrency limits."""
+        return False
+
+    @abstractmethod
+    def allocate_concurrency_slots(self, concurrency_key: str, num_int: int) -> None:
+        """Allocate concurrency slots for the given concurrency key."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_concurrency_limited_keys(self) -> Set[str]:
+        """Get the set of concurrency limited keys."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_concurrency_info(self, concurrency_key: str, include_deleted: bool = False):
+        """Get concurrency info for key."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def claim_concurrency_slot(
+        self, concurrency_key: str, run_id: str, step_key: str, priority: Optional[int] = None
+    ):
+        """Claim concurrency slots for step."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def free_concurrency_slots(self, run_id: str, step_key: Optional[str] = None):
+        """Frees concurrency slots for a given run/step."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def can_claim_from_pending(self, concurrency_key: str, run_id: str, step_key: str):
+        """Indicates that a given step can now claim a concurrency, given its priority relative to
+        other pending steps
+        .
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def remove_pending_step(self, run_id: str, step_key: Optional[str] = None):
```

Comment:
similar to above, does this deserve its own entry point? This should only happen in 
1. succesful transfer to concurrency slot
2. abandon everything for run 
?

Reply:
Initially, I had it the way you described, where claiming removed a step from `pending_steps`.  But the math / locking logic got too complicated across tables, so now there is a row in `pending_steps` for all in flight steps including ones that are currently occupying a slot.  This means that every executing step should occupy a row in the slots table and the pending steps table.

This is the state progression for steps that want to claim a slot:
1.  Step is waiting for a slot to open up (unassigned).  This state is skipped if there are more open slots than pending steps (e.g. the pending step is assigned immediately on write).
2.  A slot has opened up and the highest priority step is tapped to claim it but has not yet claimed it (assigned).
3.  The assigned step has woken up from sleep in the executor and has claimed an open slot.
4. The step has completed so the slot row is cleared and the pending step row is removed.

Reply:
But yeah, maybe we don't need this entry point...  It's called with:
1. free up slot
2. abandon everything

I initially exposed this to clean up issues where the executor hits an error, but probably we can just abandon everything and not care that we already know it has only added a pending step (not a slot).